### PR TITLE
[WIP] Add a chat example for zkSummit

### PR DESCRIPTION
**Pull request by tomaka**
~~Work in progress because I need to add support for stdin in the emscripten version of the example (the text of the `<input>` should be sent to stdin so that it's sent over the network).~~

This PR also does a bunch of bugfixes and improvements:

- Adds logging to various parts of the code (was very useful to debug the example).
- Fixes multiplexing incoming substreams being considered as being dialed instead of listened when it comes to protocol negotiation.
- The identify transport now tries to find an existing peer ID when a node connects, instead of identifying all the time. This is a critical fix, otherwise nodes would just dial back and forth to identify each other.
- The FIND_NODE kademlia RPC request now returns correct values.

### CODE REVIEW

**Comment by tomaka** [comment]
Should be good to go!

File: Cargo.toml [hunk 1]
```diff
@@ -31,3 +32,4 @@ members = [
 # Using a version of rust-multihash that compiles on emscripten
 # TODO: remove once merged upstream
 "multihash" = { git = "https://github.com/tomaka/rust-multihash", branch = "emscripten-hack" }
+"tokio-timer" = { git = "https://github.com/tomaka/tokio-timer", branch = "emscripten-test" }
```
Comment (tomaka) [inline Cargo.toml]: This contains https://github.com/tokio-rs/tokio-timer/pull/41 plus a small hack.

**Comment by tomaka** [comment]
I'm also fine with just leaving it my local repo and branch.

**Comment by tomaka** [comment]
I suggest not to merge this, this needs some polishing.

File: zkchat-example/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+        info!("Now listening on {}", actual_addr);
+        listened_addrs.write().unwrap().push(actual_addr); // TODO: wrong, requires nat_traversal
+    }
+
+    let (kad_controller, kad_init) = kad_ctl_proto.start(swarm_controller.clone());
+
+    let topic = libp2p_floodsub::TopicBuilder::new("chat").build();
+
+    let floodsub_ctl =
+        libp2p_floodsub::FloodSubController::new(&floodsub_upgrade, swarm_controller.clone());
```
Comment (quake) [inline zkchat-example/src/main.rs]: Why do we need swarm controller for floodsub? can we omit it in FloodSubController struct?

Comment (tomaka) [reply zkchat-example/src/main.rs]: True! That's a mistake.
Since this concerns a code already merged, could you open an issue instead?

Comment (quake) [reply zkchat-example/src/main.rs]: @tomaka created a PR: https://github.com/libp2p/rust-libp2p/pull/160 , please review, thanks.

**Comment by Vurich** [comment]
Are you ever going to merge this or can we nuke it?

**Comment by Vurich** [comment]
Are you ever going to merge this or can we nuke it?

**Comment by tomaka** [comment]
@Vurich Ideally this would be a full example of a chat application of libp2p. In practice this needs some more work and we keep making changes to the API.

**Comment by tomaka** [comment]
Closing. I think this chat should be rewritten at some point, but from scratch.

**Comment by tomaka** [comment]
I restored the branch, as it seems to be used by another project.

**Pull request closed without merging**